dwave/optimization/src/nodes/lambda.cpp

```diff
@@ -0,0 +1,324 @@
+// Copyright 2024 D-Wave Inc.
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include "dwave-optimization/nodes/lambda.hpp"
+
+#include "_state.hpp"
+#include "dwave-optimization/array.hpp"
+#include "dwave-optimization/nodes/inputs.hpp"
+#include "dwave-optimization/state.hpp"
+#include "dwave-optimization/utils.hpp"
+
+namespace dwave::optimization {
+
+class AccumulateZipNodeData : public ArrayNodeStateData {
+ public:
+    explicit AccumulateZipNodeData(std::vector<double>&& values,
+                                std::vector<Array::const_iterator>&& iterators, State&& state)
+            : ArrayNodeStateData(std::move(values)),
+              iterators(std::move(iterators)),
+              register_(std::move(state)) {}
+
+    // used to avoid reallocating memory for predecessor iterators every propagation
+    std::vector<Array::const_iterator> iterators;
```

Comment:
It's not in general safe to re-use iterators between propagations. Nodes might reallocate their state memory (e.g. if they're storing them in a `vector`). Because we require fixed-size predecessors, it's probably safe in practice, but not sure we should rely on that behavior.

Reply:
They are not re-used between propagations, this is just a bit of premature optimization to avoid reallocating this vector each time